SwiftCarousel.swift

```diff
@@ -221,37 +222,29 @@ public class SwiftCarousel: UIView {
         }
     }
     
-    override public func observeValueForKeyPath(keyPath: String?,
-        ofObject object: AnyObject?,
-        change: [String : AnyObject]?,
-        context: UnsafeMutablePointer<Void>) {
-            if let _ = change?[NSKeyValueChangeNewKey] where keyPath == "contentOffset" {
-                let newOffset = scrollView.contentOffset
-                if shouldScroll {
-                    if case (shouldScrollToPosition - 10..<shouldScrollToPosition + 10) = newOffset.x {
-                        shouldScroll = false
-                    }
-                }
-                
-                if !shouldScroll {
-                    var newOffsetX: CGFloat!
-                    
-                    if (newOffset.x >= scrollView.contentSize.width * 2.0 / 3.0) {
-                        newOffsetX = newOffset.x - scrollView.contentSize.width / 3.0
-                    } else if (CGRectGetMaxX(scrollView.bounds) <= scrollView.contentSize.width / 3.0) { // First part
-                        newOffsetX = newOffset.x + scrollView.contentSize.width * 3.0
-                    }
-                    
-                    guard newOffsetX != nil && newOffsetX > 0 else { return }
-                    
-                    dispatch_async(dispatch_get_main_queue(), { () -> Void in
-                        self.shouldScroll = true
-                        self.shouldScrollToPosition = newOffsetX
-                        self.scrollView.contentOffset.x = newOffsetX
-                        self.delegate?.didScroll?(toOffset: self.scrollView.contentOffset)
-                    })
-                }
+    override public func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
+        if let _ = change?[NSKeyValueChangeNewKey] where keyPath == "contentOffset" {
+            // with autolayout this seems to be quite usual, we want to wait
+            // until we have some size we can actualy work with
+            guard (scrollView.frame.width > 0 &&
+                scrollView.frame.height > 0)  else { return }
+            
+            let newOffset = scrollView.contentOffset
+            let segmentWidth = scrollView.contentSize.width / 3
+            var newOffsetX: CGFloat!
+            if (newOffset.x >= segmentWidth * 2.0) { // in the 3rd part
+                newOffsetX = newOffset.x - segmentWidth // move back one segment
+            } else if (newOffset.x + scrollView.bounds.width) <= segmentWidth { // First part
+                newOffsetX = newOffset.x + segmentWidth // move forward one segment
```

Comment:
some simplification around the math, I think there was also a bug in line `newOffsetX = newOffset.x + scrollView.contentSize.width * 3.0` and it should have been probably `newOffsetX = newOffset.x + scrollView.contentSize.width / 3.0`